### PR TITLE
Fixed server_relay flag preventing packets from processing on the server

### LINKED ISSUE #37996: Server_relay disabled, prevents mastersync from puppet function to call master
**Godot version:**
3.2.1.stable.official

**OS/device including version:**
Windows 10 Pro 18362.720

**Issue description:**
With the server's `server_relay` flag set to false, a `puppet` function does not properly invoke a `mastersync` function on the server, only locally. The client is using `rpc()`. From my understanding a `mastersync` function should always be called on the `master` and run locally on the caller. Why are these calls not making it to the server when being invoked within a `puppet` function?

Also, with the server's `server_relay` flag set to true, the `puppet` functions are properly invoked on the server, but it is also relayed to the other clients, which is what I want to avoid due to loads of errors, design, and security concerns. Why would a `mastersync` function call get relayed to other puppets in this case?

These behaviors do not seem to be consistent under these circumstances.

**Steps to reproduce:**
- Create a Server with `server_relay` set to false
```
func create_server():
	var host = NetworkedMultiplayerENet.new()
	host.server_relay = false
	host.create_server(DEFAULT_PORT, MAX_PLAYERS)
	get_tree().set_network_peer(host)
```
- Create a Client
```
func connect_to_server():
	var host = NetworkedMultiplayerENet.new()
	host.create_client(ip, DEFAULT_PORT)
	get_tree().set_network_peer(host)
```
- Create a `mastersync` function on the Server and Client
```
mastersync func test(var network_id, var val: int):
	print_debug("Test called from: ", network_id, " ", val)
	print_debug("Test running on: ", get_tree().get_network_unique_id(), " ", val)
```
- Create Client signal function and connect to `connected_to_server`
```
func _ready():
	get_tree().connect("connected_to_server", self, "_connected_ok")

func _connected_ok():
	rpc("test", get_tree().get_network_unique_id(), 1) # This works properly
```
- Create a `puppet` function on the Client that invokes the `mastersync` function
```
puppet func pre_start_game():
	rpc("test", get_tree().get_network_unique_id(), 2) # This does not work properly
```
- Run a Server and 2 Clients
- On the server, call `rpc()` to invoke the puppet function for all clients/puppets
```
rpc("pre_start_game")
```

**Results**
Server prints:
```
Test called from: 1149837705 1
Test running on: 1 1
Test called from: 111413756 1
Test running on: 1 1
```

Client A prints:
```
Test called from: 1149837705 1
Test running on: 1149837705 1
Test called from: 1149837705 2
Test running on: 1149837705 2
```

Client B prints:
```
Test called from: 111413756 1
Test running on: 111413756 1
Test called from: 111413756 2
Test running on: 111413756 2
```
**Minimal reproduction project:**
The provided projects were repurposed to only test the functionality of making an `rpc()` call from a `puppet` function to a `mastersync` function with `server_relay` disabled.

[Godot.3.2.1-puppet-mastersync-test.zip](https://github.com/godotengine/godot/files/4497612/Godot.3.2.1-puppet-mastersync-test.zip)

### PR DESCRIPTION
**Pull request by burnzy1148**
Fixes: #37996

### CODE REVIEW

File: modules/enet/networked_multiplayer_enet.cpp [hunk 1]
```diff
@@ -401,28 +405,30 @@ void NetworkedMultiplayerENet::poll() {
...
 
-								ENetPacket *packet2 = enet_packet_create(packet.packet->data, packet.packet->dataLength, packet.packet->flags);
+									ENetPacket *packet2 = enet_packet_create(packet.packet->data, packet.packet->dataLength, packet.packet->flags);
 
-								enet_peer_send(E->get(), event.channelID, packet2);
+									enet_peer_send(E->get(), event.channelID, packet2);
+								}
 							}
-
+
```
Comment (akien-mga) [inline modules/enet/networked_multiplayer_enet.cpp:422]: There is trailing whitespace on this line.

Comment (ThaBalla1148) [reply modules/enet/networked_multiplayer_enet.cpp]: Fixed

**Comment by ThaBalla1148** [comment]
If this PR is approved and merged, what are the chances of it getting cherry picked into 3.2.x?

**Comment by akien-mga** [comment]
Could you squash commits into one? See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase).

> If this PR is approved and merged, what are the chances of it getting cherry picked into 3.2.x?

That's up to @Faless, if he thinks the change is good and safe for the stable branch, we'll cherry-pick it.

**Comment by ThaBalla1148** [comment]
> 
> 
> Could you squash commits into one? See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase).
> 
> > If this PR is approved and merged, what are the chances of it getting cherry picked into 3.2.x?
> 
> That's up to @Faless, if he thinks the change is good and safe for the stable branch, we'll cherry-pick it.

Thanks. I've squashed the commits.

**Comment by ThaBalla1148** [comment]
@akien-mga Could you tell me what I need to fix in order for the build to succeed? I'm not exactly sure of the reason for failure.

**Comment by akien-mga** [comment]
There are two style issues outlined in the build log with a diff: https://travis-ci.org/github/godotengine/godot/jobs/677305686

The first one is a trailing tab on line 409, and the second is a wrongly aligned comment on line 430 (at least as far as our `clang-format` style guide is concerned).

See https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html for details about the tools we use to ensure a consistent code style.

**Comment by ThaBalla1148** [comment]
> 
> 
> There are two style issues outlined in the build log with a diff: https://travis-ci.org/github/godotengine/godot/jobs/677305686
> 
> The first one is a trailing tab on line 409, and the second is a wrongly aligned comment on line 430 (at least as far as our `clang-format` style guide is concerned).
> 
> See https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html for details about the tools we use to ensure a consistent code style.

Thank you. I've not written any C++ in over 7 years. I believe that I've fixed the remaining style issues.

**Comment by jordo** [comment]
@fales This isn't necessary if clients call set_target_peer(1) (TARGET_PEER_SERVER) before sending data through an enet channel.  The question is, do you want to support processing this message on the enet server when the target_peer is zero (TARGET_PEER_BROADCAST) when server_relay is configured to false.  I think it's reasonable to assume that TARGET_PEER_BROADCAST should be processed on server peer when server_relay is false.

**Comment by jordo** [comment]
Also, I can an backport this to 3.X if accepted.

**Comment by Faless** [comment]
Simpler solution for this PR (do we want to also handle `-N` by the same logic? I guess so).

```cpp
diff --git a/modules/enet/networked_multiplayer_enet.cpp b/modules/enet/networked_multiplayer_enet.cpp
index 25b87145b6..b2ddfb076c 100644
--- a/modules/enet/networked_multiplayer_enet.cpp
+++ b/modules/enet/networked_multiplayer_enet.cpp
@@ -371,7 +371,10 @@ void NetworkedMultiplayerENet::poll() {
                                                        // To myself and only myself
                                                        incoming_packets.push_back(packet);
                                                } else if (!server_relay) {
-                                                       // No other destination is allowed when server is not relaying
+                                                       // Other destinations will only be processed by the server when not relaying.
+                                                       if (target == 0 || (target < 0 && target != -1)) {
+                                                               incoming_packets.push_back(packet);
+                                                       }
                                                        continue;
                                                } else if (target == 0) {
                                                        // Re-send to everyone but sender :|
```

The docs should also be updated with a note in [`server_relay`](https://docs.godotengine.org/en/stable/classes/class_networkedmultiplayerenet.html#class-networkedmultiplayerenet-property-server-relay) mentioning the differnt behaviour of the `TARGET_PEER_BROADCAST` between client and server when relaying is disabled, along with the effect on peers exclusion using negative numbers.

**Comment by Faless** [comment]
> Also, I can an backport this to 3.X if accepted.

That is potentially compat breaking if someone had leftover `rpc` calls in their client code. But I guess it could be okay for `3.4`.

**Comment by jordo** [comment]
@Faless Yes i was going to mention, an cleaner solution would just be to add logic around that conditional to keep clarity.  I agree with a 3.4 merge.

**Comment by akien-mga** [comment]
Superseded by #48331. Thanks for your contribution anyway!

**Pull request closed without merging**